README.md
# belt_buffer
A buffer for your belts

A fixed-up and maintained version of klonan's belt buffer mod (https://github.com/Klonan/belt_buffer) which he seems to have abandoned.

Works fine on v18 with a bit of side touchups/improvements like a mod setting that lets you customize the buffer capacity.
PRs and forks are welcome.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.5
Date: 2020.6.28
  Changes:
    - rewrote to use the event handler
  Added:
    - new mod setting for buffer capacity
  Fixes:
    - whatever else that was broken

data.lua
local util = require "util"

local tint = function(sprite, tint)
	--local tint = tint or {}
	local sprite = util.table.deepcopy(sprite)
	sprite.tint = tint
	if sprite.hr_version then
		sprite.hr_version.tint = tint
	end
	return sprite
end

local INVENTORY_SIZE = settings.startup["belt-buffer-capacity"].value

data:extend(
	{
		{
			type = "loader",
			name = "belt-buffing-loader",
			icon = "__belt_buffer__/buffer-icon.png",
			icon_size = 32,
			flags = {"placeable-neutral", "player-creation", "fast-replaceable-no-build-while-moving", "placeable-off-grid"},
			max_health = 100,
			filter_count = 5,
			corpse = "small-remnants",
			collision_box = {{-0.2, -0.01}, {0.2, 0.01}},
			selection_box = {{-0.2, -0.2}, {0.2, 0.2}},
			selectable_in_game = false,
			animation_speed_coefficient = 32/2,
			belt_horizontal = tint(basic_belt_horizontal),
			belt_vertical = tint(basic_belt_vertical),
			ending_top = tint(basic_belt_ending_top),
			ending_bottom = tint(basic_belt_ending_bottom),
			ending_side = tint(basic_belt_ending_side),
			starting_top = tint(basic_belt_starting_top),
			starting_bottom = tint(basic_belt_starting_bottom),
			starting_side = tint(basic_belt_starting_side),
			fast_replaceable_group = "transport-belt",
			speed = 0.03125*4,
			structure =
			{
				direction_in =
				{
					sheet =
					{
					  filename = "__base__/graphics/entity/loader/loader-structure.png",
					  priority = "extra-high",
					  width = 1,
					  height = 1
					}
				},
				direction_out =
				{
					sheet =
					{
					  filename = "__base__/graphics/entity/loader/loader-structure.png",
					  priority = "extra-high",
					  width = 1,
					  height = 1
					}
				}
			},
			ending_patch = ending_patch_prototype,
			belt_distance = -0.45,
			container_distance = 0.75,
			belt_length = 0.5,
			localised_name = {"belt-buffer"},
			localised_description = {"belt-buffer-description"},
			order = "z-buffing-loader"
		}
	}
)

data:extend(
	{
		{
			type = "container",
			name = "belt-buffer-vu",
			icon = "__belt_buffer__/buffer-icon.png",
			icon_size = 32,
			flags = {"placeable-neutral", "player-creation"},
			minable = {mining_time = 1, result = "belt-buffer-proxy"},
			max_health = 150,
			corpse = "small-remnants",
			open_sound = { filename = "__base__/sound/metallic-chest-open.ogg", volume=0.65 },
			close_sound = { filename = "__base__/sound/metallic-chest-close.ogg", volume = 0.7 },
			fast_replaceable_group = "transport-belt",
			resistances =
			{
				{
					type = "fire",
					percent = 90
				},
				{
					type = "impact",
					percent = 60
				}
			},
			collision_box = {{-0.25, -0.75}, {0.25, 0.75}},
			selection_box = {{-0.5, -1}, {0.5, 1}},
			fast_replaceable_group = "container",
			inventory_size = INVENTORY_SIZE,
			vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
			picture =
			{
				filename = "__belt_buffer__/buffer-v.png",
				priority = "extra-high",
				width = 38,
				height = 64,
				shift = util.by_pixel(3, -3)
			},
			circuit_wire_connection_point = circuit_connector_definitions["chest"].points,
			circuit_connector_sprites = circuit_connector_definitions["chest"].sprites,
			circuit_wire_max_distance = default_circuit_wire_max_distance,
			localised_name = {"belt-buffer"},
			localised_description = {"belt-buffer-description"},
			order = "z-belt-buffer-chest-v"
		},
		{
			type = "container",
			name = "belt-buffer-vd",
			icon = "__belt_buffer__/buffer-icon.png",
			icon_size = 32,
			flags = {"placeable-neutral", "player-creation"},
			minable = {mining_time = 1, result = "belt-buffer-proxy"},
			max_health = 150,
			corpse = "small-remnants",
			open_sound = { filename = "__base__/sound/metallic-chest-open.ogg", volume=0.65 },
			close_sound = { filename = "__base__/sound/metallic-chest-close.ogg", volume = 0.7 },
			fast_replaceable_group = "transport-belt",
			resistances =
			{
				{
					type = "fire",
					percent = 90
				},
				{
					type = "impact",
					percent = 60
				}
			},
			collision_box = {{-0.25, -0.75}, {0.25, 0.75}},
			selection_box = {{-0.5, -1}, {0.5, 1}},
			fast_replaceable_group = "container",
			inventory_size = INVENTORY_SIZE,
			vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
			picture =
			{
				filename = "__belt_buffer__/buffer-v.png",
				priority = "extra-high",
				width = 38,
				height = 64,
				shift = util.by_pixel(3, -3)
			},
			circuit_wire_connection_point = circuit_connector_definitions["chest"].points,
			circuit_connector_sprites = circuit_connector_definitions["chest"].sprites,
			circuit_wire_max_distance = default_circuit_wire_max_distance,
			localised_name = {"belt-buffer"},
			localised_description = {"belt-buffer-description"},
			order = "z-belt-buffer-chest-v"
		},
		{
			type = "container",
			name = "belt-buffer-hl",
			icon = "__belt_buffer__/buffer-icon.png",
			icon_size = 32,
			flags = {"placeable-neutral", "player-creation"},
			minable = {mining_time = 1, result = "belt-buffer-proxy"},
			max_health = 150,
			corpse = "small-remnants",
			open_sound = { filename = "__base__/sound/metallic-chest-open.ogg", volume=0.65 },
			close_sound = { filename = "__base__/sound/metallic-chest-close.ogg", volume = 0.7 },
			fast_replaceable_group = "transport-belt",
			resistances =
			{
				{
					type = "fire",
					percent = 90
				},
				{
					type = "impact",
					percent = 60
				}
			},
			collision_box = {{-0.75, -0.25}, {0.75, 0.25}},
			selection_box = {{-1, -0.5}, {1, 0.5}},
			fast_replaceable_group = "container",
			inventory_size = INVENTORY_SIZE,
			vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
			picture =
			{
				filename = "__belt_buffer__/buffer-h.png",
				priority = "extra-high",
				width = 72,
				height = 32,
				shift = util.by_pixel(3, -3)
			},
			circuit_wire_connection_point = circuit_connector_definitions["chest"].points,
			circuit_connector_sprites = circuit_connector_definitions["chest"].sprites,
			circuit_wire_max_distance = default_circuit_wire_max_distance,
			localised_name = {"belt-buffer"},
			localised_description = {"belt-buffer-description"},
			order = "z-belt-buffer-chest-h"
		},
		{
			type = "container",
			name = "belt-buffer-hr",
			icon = "__belt_buffer__/buffer-icon.png",
			icon_size = 32,
			flags = {"placeable-neutral", "player-creation"},
			minable = {mining_time = 1, result = "belt-buffer-proxy"},
			localised_name = {"belt-buffer"},
			localised_description = {"belt-buffer-description"},
			max_health = 150,
			corpse = "small-remnants",
			open_sound = { filename = "__base__/sound/metallic-chest-open.ogg", volume=0.65 },
			close_sound = { filename = "__base__/sound/metallic-chest-close.ogg", volume = 0.7 },
			fast_replaceable_group = "transport-belt",
			resistances =
			{
				{
					type = "fire",
					percent = 90
				},
				{
					type = "impact",
					percent = 60
				}
			},
			collision_box = {{-0.75, -0.25}, {0.75, 0.25}},
			selection_box = {{-1, -0.5}, {1, 0.5}},
			fast_replaceable_group = "container",
			inventory_size = INVENTORY_SIZE,
			vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
			picture =
			{
				filename = "__belt_buffer__/buffer-h.png",
				priority = "extra-high",
				width = 72,
				height = 32,
				shift = util.by_pixel(3, -3)
			},
			circuit_wire_connection_point = circuit_connector_definitions["chest"].points,
			circuit_connector_sprites = circuit_connector_definitions["chest"].sprites,
			circuit_wire_max_distance = default_circuit_wire_max_distance,
			order = "z-belt-buffer-chest-h"
		},
	}
)

local item = util.table.deepcopy(data.raw.item["transport-belt"])
item.name = "belt-buffer-proxy"
item.place_result = "belt-buffer-proxy"
item.icon = "__belt_buffer__/buffer-icon.png"
item.icon_size = 32
item.localised_name = {"belt-buffer"}
item.localised_description = {"belt-buffer-description"}
item.order = "z-belt-buffer"
item.fast_replaceable_group = "transport-belt"
data:extend({item})

local item = util.table.deepcopy(data.raw.item["transport-belt"])
item.name = "belt-buffer-proxy-hl"
item.flags = {"hidden"}
item.place_result = "belt-buffer-hl"
item.icon = "__belt_buffer__/buffer-icon.png"
item.icon_size = 32
item.localised_name = {"belt-buffer"}
item.localised_description = {"belt-buffer-description"}
item.fast_replaceable_group = "transport-belt"
data:extend({item})

local item = util.table.deepcopy(data.raw.item["transport-belt"])
item.name = "belt-buffer-proxy-vu"
item.flags = {"hidden"}
item.place_result = "belt-buffer-vu"
item.icon = "__belt_buffer__/buffer-icon.png"
item.icon_size = 32
item.localised_name = {"belt-buffer"}
item.localised_description = {"belt-buffer-description"}
item.fast_replaceable_group = "transport-belt"
data:extend({item})

local item = util.table.deepcopy(data.raw.item["transport-belt"])
item.name = "belt-buffer-proxy-hr"
item.flags = {"hidden"}
item.place_result = "belt-buffer-hr"
item.icon = "__belt_buffer__/buffer-icon.png"
item.icon_size = 32
item.localised_name = {"belt-buffer"}
item.localised_description = {"belt-buffer-description"}
item.fast_replaceable_group = "transport-belt"
data:extend({item})

local item = util.table.deepcopy(data.raw.item["transport-belt"])
item.name = "belt-buffer-proxy-vd"
item.flags = {"hidden"}
item.place_result = "belt-buffer-vd"
item.icon = "__belt_buffer__/buffer-icon.png"
item.icon_size = 32
item.localised_name = {"belt-buffer"}
item.localised_description = {"belt-buffer-description"}
item.fast_replaceable_group = "transport-belt"
data:extend({item})

local combinator = util.table.deepcopy(data.raw["decider-combinator"]["decider-combinator"])
combinator.name = "belt-buffer-proxy"
combinator.icon = "__belt_buffer__/buffer-icon.png"
combinator.icon_size = 32
combinator.localised_name = {"belt-buffer"}
combinator.localised_description = {"belt-buffer-description"}
combinator.active_energy_usage = "1KW"
combinator.fast_replaceable_group = "transport-belt"
combinator.sprites.east = 
{
	filename = "__belt_buffer__/buffer-h.png",
	priority = "extra-high",
	width = 72,
	height = 32,
	shift = util.by_pixel(3, -3)
}
combinator.sprites.west = 
{
	filename = "__belt_buffer__/buffer-h.png",
	priority = "extra-high",
	width = 72,
	height = 32,
	shift = util.by_pixel(3, -3)
}
combinator.sprites.north = 
{
	filename = "__belt_buffer__/buffer-v.png",
	priority = "extra-high",
	width = 38,
	height = 64,
	shift = util.by_pixel(3, -3)
}
combinator.sprites.south = 
{
	filename = "__belt_buffer__/buffer-v.png",
	priority = "extra-high",
	width = 38,
	height = 64,
	shift = util.by_pixel(3, -3)
}
combinator.less_symbol_sprites.east.width = 1
combinator.less_symbol_sprites.east.height = 1
combinator.less_symbol_sprites.north.width = 1
combinator.less_symbol_sprites.north.height = 1
combinator.less_symbol_sprites.west.width = 1
combinator.less_symbol_sprites.west.height = 1
combinator.less_symbol_sprites.south.width = 1
combinator.less_symbol_sprites.south.height = 1
--error(serpent.block(combinator))
data:extend({combinator})

--- RECIPE ---

local recipe = util.table.deepcopy(data.raw.recipe["transport-belt"])
recipe.name = "belt-buffer"
recipe.result = "belt-buffer-proxy"
recipe.ingredients = 
{
	{type = "item", name = "iron-plate", amount = 5},
	{type = "item", name = "iron-gear-wheel", amount = 10},
	{type = "item", name = "transport-belt", amount = 2},
	{type = "item", name = "iron-stick", amount = 5}
}
recipe.localised_name = {"belt-buffer"}
recipe.localised_description = {"belt-buffer-description"}
recipe.result_count = 1
data:extend({recipe})

scripts/belt_buffer.lua
local util = require("util")
local buffer_enum = {
	["belt-buffer-vu"] = defines.direction.north,
	["belt-buffer-vd"] = defines.direction.south,
	["belt-buffer-hr"] = defines.direction.east,
	["belt-buffer-hl"] = defines.direction.west
}

local LOADER_ADJUSTMENT = 0.25
local SPLITTER_AREA_ADJUSTMENT = 2

---- debug method
local function stringify(var)
	local retval = ''
	if (var == nil) then
		retval = retval .. "null"
	elseif (type(var) == "table") then
		retval = retval .. serpent.block(var)
	else
		retval = retval .. tostring(var)
	end
	return retval
end
-----

built_entity = function(event)
	local entity = event.created_entity
	if not (entity and entity.valid) then return end
--		game.print("Entity built: "..stringify(entity.name))
	if entity.name == "belt-buffer-proxy" then
		local position = {x = entity.position.x, y = entity.position.y}
		local direction = entity.direction
		local surface = entity.surface
		local force = entity.force
		local definitions = entity.circuit_connection_definitions
		entity.destroy()
		--error(serpent.block(definitions))
--			game.print(direction)
			-- 0 is bottom-to-top
			-- 2 is left-to-right
			-- 4 is top-to-bottom
			-- 6 is right-to-left
		local loader_in
		local loader_out
		if direction == 0 then
			loader_in = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x, y = position.y + LOADER_ADJUSTMENT},
				direction = 0,
				force = force
			}
			loader_out = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x, y = position.y - LOADER_ADJUSTMENT},
				direction = 4,
				force = force
			}
		elseif direction == 2 then
			loader_in = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x - LOADER_ADJUSTMENT, y = position.y},
				direction = 2,
				force = force
			}
			loader_out = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x + LOADER_ADJUSTMENT, y = position.y},
				direction = 6,
				force = force
			}
		elseif direction == 4 then
			loader_in = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x, y = position.y - LOADER_ADJUSTMENT},
				direction = 4,
				force = force
			}
			loader_out = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x, y = position.y + LOADER_ADJUSTMENT},
				direction = 0,
				force = force
			}
		elseif direction == 6 then
			loader_in = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x + LOADER_ADJUSTMENT, y = position.y},
				direction = 6,
				force = force
			}
			loader_out = surface.create_entity{
				name = "belt-buffing-loader",
				position = {x = position.x - LOADER_ADJUSTMENT, y = position.y},
				direction = 2,
				force = force
			}
		end
--		game.print("loader in: "..stringify(loader_in).." is valid? "..stringify(loader_in.valid))
--		game.print("loader out: "..stringify(loader_out).." is valid? "..stringify(loader_out.valid))
		if not (loader_out and loader_out.valid and loader_in and loader_in.valid) then
			--one of them was fucked, cancel the whole thing
--			game.print("Something went wrong, check debug logs to figure out what")
			if loader_out then
				loader_out.destroy()
			end
			if loader_in then
				loader_in.destroy()
			end
			if event.player_index then
				game.players[event.player_index].insert({name = "belt-buffer-proxy"})
			else
				local stack = surface.create_entity{name = "item-entity", stack = {name = "belt-buffer-proxy", count = 1}, position = position, force = force}
				stack.order_deconstruction(force)
			end
			return
		end
		loader_out.loader_type = "output"
		loader_in.destructible = false
		loader_out.destructible = false
		local chest
		if direction == 0	then
			chest = surface.create_entity{name = "belt-buffer-vu", position = position, force = force}
		elseif direction == 4 then
			chest = surface.create_entity{name = "belt-buffer-vd", position = position, force = force}
		elseif direction == 2 then
			chest = surface.create_entity{name = "belt-buffer-hr", position = position, force = force}
		elseif direction == 6 then
			chest = surface.create_entity{name = "belt-buffer-hl", position = position, force = force}
		end
		for k, connection in pairs (definitions) do
			chest.connect_neighbour(connection)
		end
		return
	end
	if entity.name == "entity-ghost" then
--			game.print("entity is ghost")
		local position = {x = entity.position.x, y = entity.position.y}
		local direction = entity.direction
		local surface = entity.surface
		local force = entity.force
		local inner_entity = entity.ghost_name
		local direction = buffer_enum[inner_entity]
		if direction then
			entity.destroy()
			surface.create_entity{name = "entity-ghost", inner_name = "belt-buffer-proxy", direction = direction, force = force, position = position, expires = false}
		end
	end
end

mined_entity = function(event)
	local entity = event.entity
	if not (entity and entity.valid) then return end
	remove_nearby_loaders(entity.name, entity.position, entity.surface)
end

remove_nearby_loaders = function(name, position, surface)
	if name == "belt-buffer-vu" or name == "belt-buffer-vd" then
		local p1 = {x = position.x, y = position.y - LOADER_ADJUSTMENT}
		local splitter_area = {{p1.x - SPLITTER_AREA_ADJUSTMENT, p1.y - SPLITTER_AREA_ADJUSTMENT},{p1.x + SPLITTER_AREA_ADJUSTMENT, p1.y + SPLITTER_AREA_ADJUSTMENT}}
		local splitters = surface.find_entities_filtered{area = splitter_area, type = "splitter"}
		local area = {{p1.x - 0.01, p1.y - 0.01},{p1.x + 0.01, p1.y + 0.01}}
		local nearby = surface.find_entities_filtered{area = area, name = "belt-buffing-loader"}
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		for k, loader in pairs (nearby) do
			loader.destructible = true
			loader.destroy()
--				game.print("killed him")
			break
		end
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		local p2 = {x = position.x, y = position.y + LOADER_ADJUSTMENT}
		splitter_area = {{p2.x - SPLITTER_AREA_ADJUSTMENT, p2.y - SPLITTER_AREA_ADJUSTMENT},{p2.x + SPLITTER_AREA_ADJUSTMENT, p2.y + SPLITTER_AREA_ADJUSTMENT}}
		local splitters = surface.find_entities_filtered{area = splitter_area, type = "splitter"}
		local area = {{p2.x - 0.01, p2.y-0.01},{p2.x+0.01, p2.y+0.01}}
		local nearby = surface.find_entities_filtered{area = area, name = "belt-buffing-loader"}
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		for k, loader in pairs (nearby) do
			loader.destructible = true
			loader.destroy()
--				game.print("killed his friend")
			break
		end
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		return
	end
	if name == "belt-buffer-hl" or name == "belt-buffer-hr" then
		local p1 = {x = position.x - LOADER_ADJUSTMENT, y = position.y}
		local area = {{p1.x - 0.01, p1.y-0.01},{p1.x+0.01, p1.y+0.01}}
		local splitter_area = {{p1.x - SPLITTER_AREA_ADJUSTMENT, p1.y - SPLITTER_AREA_ADJUSTMENT},{p1.x + SPLITTER_AREA_ADJUSTMENT, p1.y + SPLITTER_AREA_ADJUSTMENT}}
		local splitters = surface.find_entities_filtered{area = splitter_area, type = "splitter"}
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		local nearby = surface.find_entities_filtered{area = area, name = "belt-buffing-loader"}
		for k, loader in pairs (nearby) do
			loader.destructible = true
			loader.destroy()
--				game.print("killed him")
			break
		end
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		local p2 = {x = position.x + LOADER_ADJUSTMENT, y = position.y}
		splitter_area = {{p2.x - SPLITTER_AREA_ADJUSTMENT, p2.y - SPLITTER_AREA_ADJUSTMENT},{p2.x + SPLITTER_AREA_ADJUSTMENT, p2.y + SPLITTER_AREA_ADJUSTMENT}}
		local splitters = surface.find_entities_filtered{area = splitter_area, type = "splitter"}
		local area = {{p2.x - 0.01, p2.y-0.01},{p2.x+0.01, p2.y+0.01}}
		local nearby = surface.find_entities_filtered{area = area, name = "belt-buffing-loader"}
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		for k, loader in pairs (nearby) do
			loader.destructible = true
			loader.destroy()
--				game.print("killed his friend")
			break
		end
		for k, splitter in pairs (splitters) do
			if splitter.valid then
				splitter.rotate()
			end
		end
		return
	end
end


function on_player_pipette(event)
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end
	local selected = player.selected
	local item = event.item
	if item and buffer_enum[item.name] then
		player.remove_item(item.name)
		player.cursor_stack.set_stack{name = "belt-buffer-proxy", count = game.item_prototypes["belt-buffer-proxy"].stack_size}
	end
	if selected and selected.valid and buffer_enum[selected.name] then
		local count = math.min(player.get_item_count("belt-buffer-proxy"), game.item_prototypes["belt-buffer-proxy"].stack_size)
		if player.cheat_mode then count = game.item_prototypes["belt-buffer-proxy"].stack_size end
		if count > 0 then
			player.cursor_stack.set_stack{name = "belt-buffer-proxy", count = count}
			player.remove_item{name = "belt-buffer-proxy", count = count}
		end
	end
end

function on_entity_died(event)
	local entity = event.entity
	if not (entity and entity.valid) then return end
	local direction = buffer_enum[entity.name]
	if not direction then return end
	mined_entity(event)
--		game.print("It died to")
	local position = {x = entity.position.x, y = entity.position.y}
	local surface = entity.surface
	local force = entity.force
	local prototype = entity.prototype
	local corpse_name
	for name, corpse in pairs (prototype.corpses) do
		corpse_name = name
		break
	end
	entity.destroy()
	if corpse_name then
		surface.create_entity{name = corpse_name, force = "neutral", position = position}
	end
	surface.create_entity{name = "entity-ghost", inner_name = "belt-buffer-proxy", direction = direction, force = force, position = position, expires = true}
end


function on_player_configured_blueprint(event)
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end
	local stack = player.cursor_stack
	if not (stack and stack.valid and stack.valid_for_read) then return end
--		game.print("Stack is valid")
	local entities = stack.get_blueprint_entities()
	if not entities then return end
--		error(serpent.block(entities))
	local adjust_connections = {}
	for k, entity in pairs (entities) do
		local direction = buffer_enum[entity.name]
		if direction then
			entity.name = "belt-buffer-proxy"
			entity.direction = direction
			if entity.connections then
				adjust_connections[entity.entity_number] = true
			end
		end
	end
	for k, entity in pairs (entities) do
		if entity.connections then
			for k, connection in pairs (entity.connections) do
				for k, name in pairs ({"red", "green"}) do
					if connection[name] then
						for k, hassle in pairs (connection[name]) do
							if adjust_connections[hassle.entity_id] then
								hassle.circuit_id = 1
							end
						end
					end
				end
			end
		end
	end
	--error(serpent.block(entities))
	stack.set_blueprint_entities(entities)
end

----------------------------------------------------------------------------------
-- will probably get rid of this, because when other people break this mod they --
-- should fix it to remain compatible instead od authors/maintainers doing it	--
----------------------------------------------------------------------------------
remote_conditionals = function()
	if remote.interfaces["picker"] and remote.interfaces["picker"]["dolly_moved_entity_id"] then
		script.on_event(remote.call("picker", "dolly_moved_entity_id"), nexelas_mod_moved_something_so_its_my_job_to_fix_it)
	end
end

--[[player_index = player_index, --The index of the player who moved the entity
	moved_entity = entity, --The entity that was moved
	start_pos = position --The position that the entity was moved from]]

nexelas_mod_moved_something_so_its_my_job_to_fix_it = function(event)
	local buffer = event.moved_entity
	if not (buffer and buffer.valid) then return end
	local direction =	buffer_enum[buffer.name]
	if not direction then return end
	remove_nearby_loaders(buffer.name, event.start_pos, buffer.surface)
	local position = buffer.position
	local surface = buffer.surface
	local force = buffer.force
	buffer.destroy()
	local proxy = surface.create_entity{name = "belt-buffer-proxy", position = position, direction = direction, force = force, surface = surface}
	built_entity({created_entity = proxy, player_index = event.player_index})
end



------------------
-- library part --
------------------

local lib = {}

lib.events =
{
	[defines.events.on_robot_mined_entity] = mined_entity,
	[defines.events.on_player_mined_entity] = mined_entity,
	[defines.events.on_robot_built_entity] = built_entity,
	[defines.events.on_built_entity] = built_entity,
	[defines.events.on_player_pipette] = on_player_pipette,
	[defines.events.on_entity_died] = on_entity_died,
	[defines.events.on_player_configured_blueprint] = on_player_configured_blueprint
}

lib.on_load = function()
	remote_conditionals()
end

lib.on_configuration_changed = function()

end

lib.on_init = function()
	remote_conditionals()
end

return lib